server/src/main/java/io/unitycatalog/server/persist/CredentialRepository.java

```diff
@@ -0,0 +1,235 @@
+import io.unitycatalog.server.exception.BaseException;
+import io.unitycatalog.server.exception.ErrorCode;
+import io.unitycatalog.server.model.*;
+import io.unitycatalog.server.persist.dao.CredentialDAO;
+import io.unitycatalog.server.persist.utils.PagedListingHelper;
+import io.unitycatalog.server.utils.IdentityUtils;
+import io.unitycatalog.server.utils.ValidationUtils;
+import java.time.Instant;
+import java.util.*;
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.hibernate.Transaction;
+import org.hibernate.query.Query;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CredentialRepository {
+  private static final Logger LOGGER = LoggerFactory.getLogger(CredentialRepository.class);
+  private final Repositories repositories;
+  private final SessionFactory sessionFactory;
+  private static final PagedListingHelper<CredentialDAO> LISTING_HELPER =
+      new PagedListingHelper<>(CredentialDAO.class);
+  public static ObjectMapper objectMapper = new ObjectMapper();
+
+  public CredentialRepository(Repositories repositories, SessionFactory sessionFactory) {
+    this.repositories = repositories;
+    this.sessionFactory = sessionFactory;
+  }
+
+  public CredentialInfo addCredential(CreateCredentialRequest createCredentialRequest) {
+    ValidationUtils.validateSqlObjectName(createCredentialRequest.getName());
+    String callerId = IdentityUtils.findPrincipalEmailAddress();
+    UUID storageCredentialId = UUID.randomUUID();
+    CredentialInfo storageCredentialInfo =
+        new CredentialInfo()
+            .id(storageCredentialId.toString())
+            .name(createCredentialRequest.getName())
+            .comment(createCredentialRequest.getComment())
+            .purpose(createCredentialRequest.getPurpose())
+            .owner(callerId)
+            .createdAt(Instant.now().toEpochMilli())
+            .createdBy(callerId);
+
+    if (createCredentialRequest.getAwsIamRole() != null) {
+      storageCredentialInfo.setAwsIamRole(
+          fromAwsIamRoleRequest(createCredentialRequest.getAwsIamRole()));
+    } else {
+      throw new BaseException(
+          ErrorCode.INVALID_ARGUMENT,
+          "Storage credential must have one of aws_iam_role, azure_service_principal, azure_managed_identity or gcp_service_account");
+    }
+
+    try (Session session = sessionFactory.openSession()) {
+      Transaction tx = session.beginTransaction();
+      try {
+        if (getCredentialDAO(session, createCredentialRequest.getName()) != null) {
+          throw new BaseException(
+              ErrorCode.ALREADY_EXISTS,
+              "Storage credential already exists: " + createCredentialRequest.getName());
+        }
+        session.persist(CredentialDAO.from(storageCredentialInfo));
+        LOGGER.info("Added storage credential: {}", storageCredentialInfo.getName());
+        tx.commit();
+        return storageCredentialInfo;
+      } catch (Exception e) {
+        tx.rollback();
```

Comment:
can we deduplicate this repeated logic of handling txns? i am guessing that this is duplicate across many repositories. We can start with adding a util function/common trait and using it in just this repo first.